### PR TITLE
12 implement common terms query

### PR DESCRIPTION
**Pull request by TomChaton**
Do not merge until https://github.com/tomcroft/tantrum_elastic/pull/89 has been merged

### CODE REVIEW

**Comment by carlcasbolt** [comment]
#89 has not been merged, does this need to be rebased?

File: phpunit.xml [hunk 1]
```diff
@@ -12,4 +12,4 @@
         <log type="coverage-html" target="build/logs/codeCoverage" charset="UTF-8" highlight="false" lowUpperBound="35" highLowerBound="70"/>
         <log type="coverage-clover" target="build/logs/clover.xml" />
     </logging>
-</phpunit>
+</phpunit>
```
Comment (luispabon) [inline phpunit.xml:15]: Add newline at the end of file.

File: src/Exception/IncompatibleValues.php [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace tantrum_elastic\Exception;
+
+class IncompatibleValues extends Validation
+{
+}
```
Comment (luispabon) [inline src/Exception/IncompatibleValues.php]: Add newline at the end of file.

File: src/Lib/Container.php [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace tantrum_elastic\Lib;
+
+abstract class Container extends Element
+{
+}
```
Comment (luispabon) [inline src/Lib/Container.php:7]: Add newline at the end of file.

File: src/Lib/Element.php [hunk 4]
```diff
@@ -14,185 +14,207 @@
...
-            throw new Exception\NotSupported('Cannot process options for var of type "%s"', gettype($serializable));
+        $elements = [];
+
+        // Fetch external elements from the elements
+        foreach($this->elements as $element) {
+            $elements = array_merge($elements, $element->getExternalElements());
         }
+
+        //Merge them with our elements
+        $elements = array_merge($this->elements, $elements);
```
Comment (luispabon) [inline src/Lib/Element.php:143]: Not sure I understand this... you're foreaching `$this->elements` and merging on to `$elements` before, then you merge again at the top level?

File: src/Lib/Element.php [hunk 5]
```diff
@@ -14,185 +14,207 @@
...
+    protected function preProcess()
+    {
+    }
+
+    /**
+     * Allows descendants to validate their values before processing
+     */
+    protected function validate()
+    {
+    }
```
Comment (luispabon) [inline src/Lib/Element.php:208]: Why not declare these two abstract?

File: src/Query/CommonTerms.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        $this->addElement($this->body);
+        return $this->elements;
+    }
+
+    public function getElementName()
+    {
+        return 'common';
+    }
+}
```
Comment (luispabon) [inline src/Query/CommonTerms.php]: Add newline at the end of file.

**Comment by TomChaton** [comment]
Closing this, as there are some weird intermittent unit test failures

**Pull request closed without merging**